python/hsfs/core/schema_validation.py

```diff
@@ -0,0 +1,225 @@
+
+    def get_online_varchar_length(self, feature):
+        # check of online_type is not null and starts with varchar
+        if (
+            feature
+            and feature.online_type
+            and feature.online_type.startswith("varchar")
+        ):
+            return int(self.extract_numbers(feature.online_type)[0])
+        return None
+
+    @staticmethod
+    def increase_string_columns(column_lengths: dict, dataframe_features):
+        # Adjusts the length of varchar columns
+        for i_feature in dataframe_features:
+            if i_feature.name in column_lengths:
+                i_feature.online_type = f"varchar({column_lengths[i_feature.name]})"
+                logging.warning(
+                    f"Column {i_feature.name} maximum string length increased to {column_lengths[i_feature.name]}"
+                )
+        return dataframe_features
+
+
+class PandasValidator(DataFrameValidator):
+    # Pandas df specific validator
+    def _validate_df_specifics(self, feature_group, df):
+        errors = {}
+        column_lengths = {}
+        is_pk_null = False
+        is_string_length_exceeded = False
+
+        # Check for null values in primary key columns
+        for pk in feature_group.primary_key:
+            if df[pk].isnull().any():
+                errors[pk] = f"Primary key column {pk} contains null values."
+                is_pk_null = True
+
+        # Check string lengths
+        for col in df.select_dtypes(include=["object"]).columns:
+            currentmax = df[col].str.len().max()
+            col_max_len = (
+                self.get_online_varchar_length(
+                    self.get_feature_from_list(col, feature_group.features)
+                )
+                if feature_group.features
+                else 100
+            )
+
+            if col_max_len is not None and currentmax > col_max_len:
+                errors[col] = (
+                    f"String length exceeded. Column {col} has string values longer than maximum column limit of {col_max_len} characters."
+                )
+                column_lengths[col] = currentmax
+                is_string_length_exceeded = True
+
+        return errors, column_lengths, is_pk_null, is_string_length_exceeded
+
+
+class PolarsValidator(DataFrameValidator):
+    # Polars df specific validator
+    def _validate_df_specifics(self, feature_group, df):
+        import polars as pl
+
+        errors = {}
+        column_lengths = {}
+        is_pk_null = False
+        is_string_length_exceeded = False
+
+        # Check for null values in primary key columns
+        for pk in feature_group.primary_key:
+            if df[pk].is_null().any():
+                errors[pk] = f"Primary key column {pk} contains null values."
+                is_pk_null = True
+
+        # Check string lengths
+        for col in df.select(pl.col(pl.Utf8)).columns:
+            currentmax = df[col].str.len_chars().max()
```

Comment:
Can you confirm with RonDB team whether we should use len_bytes or len_char here? I am not sure what VARCHAR(100) is setup to use in rondb for online feature store table

Reply:
In database it allocates based on bytes. e.g. for varchar(100) it would allocated approx 400 bytes. So I think checking the len_chars works since we create the table based on the length, or also could define so in the schema.